### PR TITLE
WT-7419 Tiered local storage changes to use WT file system.

### PR DESCRIPTION
**Pull request by ddanderson**
For the local file system open, use the WT file system to get a file handle, rather than using POSIX file descriptors directly.
This makes for a cleaner implementation.  Get rid of marker files and temp files. We don't need temp files because our model is to behave exactly like a "regular" file system until flush occurs. We also don't need a flush marker, when a flush occurs, we chmod the file to readonly.  That matches the semantics - flushed files are by definition immutable.

### CODE REVIEW

File: ext/storage_sources/local_store/local_store.c [hunk 1]
```diff
@@ -729,58 +736,58 @@ local_open(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name,
...
-
-    local->op_count++;
-    if (flags == WT_FS_OPEN_CREATE)
-        oflags = O_WRONLY | O_CREAT;
-    else if (flags == WT_FS_OPEN_READONLY)
-        oflags = O_RDONLY;
-    else
-        return (local_err(local, session, EINVAL, "open: invalid flags: 0x%x", flags));
+    /*
+     * If needed we could relax this, and allow opens of other kinds of files to use the regular
```
Comment (sueloverso) [inline ext/storage_sources/local_store/local_store.c:755]: I suggest rewording this comment because "If needed we could relax this" is the first thing and I have no idea what "this" is nor what constraints you're talking about. Maybe something like:
Restrict opens to regular data files. If needed, ...

also does that mean that we could not open/flush WT log files?

File: src/include/wiredtiger_ext.h [hunk 2]
```diff
@@ -238,6 +238,21 @@ struct __wt_extension_api {
...
       WT_CONFIG_ARG *config, WT_CONFIG_PARSER **config_parserp);
 
+    /*!
+     * Get the file system abstraction used by WiredTiger.
+     *
+     * @param wt_api the extension handle
+     * @param session the session handle (or NULL if none available)
+     * @param file_system the returned file system handle.
+     * @errors
+     * If called when from an extension's initialization routine, this may
```
Comment (sueloverso) [inline src/include/wiredtiger_ext.h:248]: "If called when from" I think 'when' should be removed.

**Review by sueloverso — APPROVED** [review]
LGTM I have a few minor comments and questions.

File: ext/storage_sources/local_store/local_store.c (hunk 1 above)
Comment (ddanderson) [reply ext/storage_sources/local_store/local_store.c:755]: We normally wouldn't be using this file system to flush WT log files.  The block manager PR has code like:
```
  WT_WITH_FILE_SYSTEM(....)
```
that is used narrowly, like when creating a data file, or opening a data file or the turtle file.  It would be unexpected to try to open a log file under that umbrella, so I made it an error here so we understand it.  That said, that constraint could be relaxed pretty easily if we wanted.  I'll update the comment to give this context.

File: ext/storage_sources/local_store/local_store.c [hunk 3]
```diff
@@ -729,58 +739,65 @@ local_open(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name,
...
         goto err;
-    if (flags == WT_FS_OPEN_CREATE) {
+
+    ret = stat(local_fh->path, &sb);
+    if (ret != 0 && errno != ENOENT)
+        ret = local_err(local, session, errno, "%s: local_open stat", local_fh->path);
+    exists = (ret == 0);
+
+    /*
+     * TODO: tiered: If the file doesn't exist locally, make a copy of it from the cloud here.
```
Comment (keitharnoldsmith) [inline ext/storage_sources/local_store/local_store.c:786]: Sanity checking my thinking here. 

If the file is in the cloud and we don't have a local copy, could we just read the blocks we need from cloud, without doing the copy?

Even if we eventually want to copy the object to the local file system, it might be nice to handle an immediate read of a few KB without waiting to download a multi-MB (or multi-GB) object first.

I'm just thinking out loud. No need to change anything here.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM. 

One comment that's more of a "future design" question than anything that needs to be addressed in this PR.

File: ext/storage_sources/local_store/local_store.c (hunk 3 above)
Comment (ddanderson) [reply ext/storage_sources/local_store/local_store.c:786]: Yes, we could.  I've not (yet) implemented any removal from the local directory, but when we do, we could explore what you're talking about it.  The easiest solution is to get the whole file every time of course.  But in fact what we'll probably have a client of this fs (i.e. the block manager) that has a file handle still open, and it eventually will do a read of a block at an offset.  What we don't know is whether we'll have more reads coming, but the right thing may well be to "page in" just the one block or a chunk around it.  It would require some per-file tracking, but totally doable.
